### DISCUSSION TITLE
Any guidance on how to use await / top level await

### DISCUSSION THREAD

**Discussion by kpko** [sebastienros/jint#1607] [Q&A]
Hi,

in the long term I want to execute async c# functions in JavaScript. With the recent advantages with async/await and top level awaits I tried integrating these functions again. I understand there is no implicit task/promise conversion so I tried to use the engine.RegisterPromise() method to return a promise using a simple delay method.

As of my understanding this should create a promise that gets resolved in 5 seconds. The promise should be awaitable because of top-level awaits. 

```c#
using Jint;

var engine = new Engine();
engine.SetValue("delay", (int seconds) =>
{
    var promise = engine.RegisterPromise();
    _ = Task.Delay(seconds * 1000)
        .ContinueWith(_ => promise.Resolve("success"), TaskContinuationOptions.OnlyOnRanToCompletion)
        .ContinueWith(_ => promise.Reject("failed"), TaskContinuationOptions.OnlyOnFaulted);

    return promise.Promise;
});
var val = engine.Evaluate("await delay(5)");
Console.WriteLine(val.UnwrapIfPromise());
```

This errors with 'Unexpected token' though. I tried wrapping everything in an async method:

```c#
var val = engine.Evaluate("(async () => await delay(5))()");
Console.WriteLine(val.UnwrapIfPromise());
```

But then I get the error 'System.InvalidOperationException: 'UnwrapIfPromise' called before Promise was settled'. This error seems to occur in the .Evaluate() line so I can't even work around that using a task completion source or something.

Is it possible to use await to await a task (converted to promise) in a script at this point? Or is what I'm trying simply not possible now (or ever)? Is there something I am missing?

Thank you in advance!

**Comment by Xicy** [comment]
I tried to implement Task to Promise conversion in that PR #1567. It is working but it's not actually async, it's block the thread until resolve. 
You can follow up in the pr. I hope PR helps to you :)

**Reply by kpko** [reply]
@lahma thanks for moving :)

@Xicy Hey! Yeah I saw that and I’m looking forward to the merge. But would this help in this scenario?

I think my problem is twofold: first getting async stuff into jint and second getting async stuff back out. Task to promise conversion and only helps with the first problem, right? I will look into the PR again. :) thank you.

**Reply by Xicy** [reply]
Code block which is below, solve your problems. SetValue move to Task method to JsPromise instance. UnWrapIfPromise is resolve the promise to actual result. (ps: I miss top level awaiting, I'm not sure is it working with the pr.)

You got the error(unsettlement) when you create a instance of promise because engine does not know promise has to be resolve. So I add the promise in eventloop(as I remember) then engine resolve waiting promises at the end of Evaluate method. 

Sorry If I miss word or miss remember of methods name, b/c I'm away from pc :) 

```cs

public void ShouldTaskConvertedToPromiseInJS()
    {
        Engine engine = new();
        engine.SetValue("callable", Callable);
        var result = engine.Evaluate("callable().then(x=>x*2)");
        result = result.UnwrapIfPromise();
        Assert.Equal(2, result);

        static async Task<int> Callable()
        {
            await Task.Delay(10);
            Assert.True(true);
            return 1;
        }
    }

```

**Reply by kpko** [reply]
I think I'm starting to understand what went wrong here. So the promise has to be "registered" in the engine somehow, so Jint knows there are things to wait for? And because this doesn't happen, the engine doesn't know the promise has to be settled first?

My question would be, what is the engine.RegisterPromise() method for if not registering a promise in the engine? What do I need to do to make the engine "aware" of the new promise that gets returned from my method here?

```csharp
var engine = new Engine();
engine.SetValue("delay", (int seconds) =>
{
    var promise = engine.RegisterPromise();
    _ = Task.Delay(seconds * 1000)
        .ContinueWith(_ => { promise.Resolve("success"); }, TaskContinuationOptions.OnlyOnRanToCompletion)
        .ContinueWith(_ => { promise.Reject("failed"); }, TaskContinuationOptions.OnlyOnFaulted);

    return promise.Promise;
});

var value = engine.Evaluate("delay(5).then(x => x)");
Console.WriteLine(value.UnwrapIfPromise());
```

**Reply by kpko** [reply]
When looking into the implementation of "RegisterPromise", I saw the comment that this is still experimental and has to be called with "ExecuteWithEventLoop", a method that I can't find (at least on the engine).

Even when I try to wait for the internal delay method, the call seems to work using a promise but fails with the unsettled promise error using an async method. 

```csharp
using Jint;

var engine = new Engine();
var tcs = new TaskCompletionSource();
engine.SetValue("delay", (int seconds) =>
{
    var promise = engine.RegisterPromise();
    _ = Task.Delay(seconds * 1000)
        .ContinueWith(_ =>
        {
            promise.Resolve("success");
            tcs.SetResult();
        }, TaskContinuationOptions.OnlyOnRanToCompletion);

    return promise.Promise;
});

// Doesn't work:
//var value = engine.Evaluate("(async () => await delay(5))().then(x => x)");

// Does work:
var value = engine.Evaluate("delay(5).then(x => x)");

await tcs.Task;
Console.WriteLine(value.UnwrapIfPromise());
```
